pkg/build/api/validation/validation.go

```diff
@@ -106,30 +106,19 @@ func ValidateBuildRequest(request *buildapi.BuildRequest) fielderrors.Validation
 	allErrs := fielderrors.ValidationErrorList{}
 	allErrs = append(allErrs, validation.ValidateObjectMeta(&request.ObjectMeta, true, oapi.MinimalNameRequirements).Prefix("metadata")...)
 
-	if request.Revision != nil {
-		allErrs = append(allErrs, validateRevision(request.Revision).Prefix("revision")...)
-	}
 	return allErrs
 }
 
 func validateBuildSpec(spec *buildapi.BuildSpec) fielderrors.ValidationErrorList {
 	allErrs := fielderrors.ValidationErrorList{}
-	hasSourceType := len(spec.Source.Type) != 0
-	switch t := spec.Strategy.Type; {
-	// 'source' is optional for Custom builds
-	case t == buildapi.CustomBuildStrategyType && hasSourceType:
-		allErrs = append(allErrs, validateSource(&spec.Source).Prefix("source")...)
-	case t == buildapi.SourceBuildStrategyType:
-		allErrs = append(allErrs, validateSource(&spec.Source).Prefix("source")...)
-		if spec.Source.Type == buildapi.BuildSourceDockerfile {
-			allErrs = append(allErrs, fielderrors.NewFieldInvalid("source.type", nil, "may not be type Dockerfile for source builds"))
-		}
-	case t == buildapi.DockerBuildStrategyType:
-		allErrs = append(allErrs, validateSource(&spec.Source).Prefix("source")...)
-	}
-	if spec.Revision != nil {
-		allErrs = append(allErrs, validateRevision(spec.Revision).Prefix("revision")...)
+	s := spec.Strategy
+
+	if s.CustomStrategy != nil && spec.Source.Git == nil && spec.Source.Binary == nil && spec.Source.Dockerfile == nil {
+		allErrs = append(allErrs, fielderrors.NewFieldInvalid("source", spec.Source, "must provide a value for at least one of source, binary, or dockerfile"))
 	}
+
+	allErrs = append(allErrs, validateSource(&spec.Source, s.CustomStrategy != nil).Prefix("source")...)
```

Comment:
that's going to make the logic incredibly complicated.  Since validateStrategy is now the only thing checking it multiple strategies are set, we'd have to check the return error from validateStrategy, decide if it as a "multiple strategies" validation error, then do what you suggest.

if validateStrategy returned some other validation error, it would still be reasonable to then call validateSource to check the source.

I don't think it's worth the headache.